File: tinysoul/endpoint/http/routes/runtime.py
"""Runtime status and command routes."""

from fastapi import FastAPI

from tinysoul.infra.json import JsonObject, to_json_object

from ...engine import EndpointControlKind, EndpointEngine
from ..schemas import ControlRequest, InputRequest


def register_runtime_routes(app: FastAPI, engine: EndpointEngine) -> None:
    @app.get("/v1/status")
    def status() -> JsonObject:
        return engine.runtime.status()

    @app.post("/v1/input", status_code=202)
    def submit_input(body: InputRequest) -> JsonObject:
        return engine.runtime.submit_user_input(
            body.text,
            to_json_object(body.metadata),
            command_id=body.command_id,
        )

    @app.post("/v1/control", status_code=202)
    def submit_control(body: ControlRequest) -> JsonObject:
        return engine.runtime.submit_control(
            EndpointControlKind(body.kind),
            to_json_object(body.metadata),
            command_id=body.command_id,
        )

File: visualization/src/api/runtime.ts
/** Runtime status and command client. */

import type {
  BackendStatus,
  CommandReceipt,
  ControlRequest,
  InputRequest,
} from "../types";
import type { EndpointHttpTransport } from "./transport";

export class EndpointRuntimeClient {
  constructor(private readonly transport: EndpointHttpTransport) {}

  status(): Promise<BackendStatus> {
    return this.transport.request<BackendStatus>("GET", "/v1/status");
  }

  submitInput(request: InputRequest): Promise<CommandReceipt> {
    return this.transport.request<CommandReceipt>("POST", "/v1/input", {
      body: request,
    });
  }

  submitControl(request: ControlRequest): Promise<CommandReceipt> {
    return this.transport.request<CommandReceipt>("POST", "/v1/control", {
      body: request,
    });
  }
}
